### PR TITLE
Implemented volume topology aware scheduling

### PR DESCRIPTION
**Pull request by ellistarn**
**1. Issue, if available:**
#622 

**2. Description of changes:**
### Dynamic Provisioning
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: ebs-sc
provisioner: ebs.csi.aws.com
volumeBindingMode: WaitForFirstConsumer
allowedTopologies:
- matchLabelExpressions:
  - key: topology.ebs.csi.aws.com/zone
    values:
    - us-west-2d
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: ebs-claim
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: ebs-sc
  resources:
    requests:
      storage: 4Gi
apiVersion: v1
kind: Pod
metadata:
  name: app
spec:
  nodeSelector:
    karpenter.sh/provisioner-name: default
  containers:
  - name: app
    image: centos
    command: ["/bin/sh"]
    args: ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"]
    volumeMounts:
    - name: persistent-storage
      mountPath: /data
  volumes:
  - name: persistent-storage
    persistentVolumeClaim:
      claimName: ebs-claim
```
### Static Provisioning
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: test-pv
spec:
  capacity:
    storage: 50Gi
  volumeMode: Filesystem
  accessModes:
    - ReadWriteOnce
  storageClassName: ""
  csi:
    driver: ebs.csi.aws.com
    volumeHandle: vol-0755eac953162225d
    fsType: ext4
  nodeAffinity:
    required:
      nodeSelectorTerms:
        - matchExpressions:
            - key: topology.ebs.csi.aws.com/zone
              operator: In
              values:
                - us-west-2d
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: ebs-claim
spec:
  accessModes:
    - ReadWriteOnce
  storageClassName: ""
  resources:
    requests:
      storage: 50Gi
---
apiVersion: v1
kind: Pod
metadata:
  name: app
spec:
  nodeSelector:
    karpenter.sh/provisioner-name: "default"
  containers:
    - name: app
      image: centos
      command: ["/bin/sh"]
      args:
        ["-c", "while true; do echo $(date -u) >> /data/out.txt; sleep 5; done"]
      volumeMounts:
        - name: persistent-storage
          mountPath: /data
  volumes:
    - name: persistent-storage
      persistentVolumeClaim:
        claimName: ebs-claim
```

**3. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/provisioning/provisioner.go [hunk 1]
```diff
@@ -249,6 +247,26 @@ func (p *Provisioner) bind(ctx context.Context, node *v1.Node, pods []*v1.Pod) (
...
+	}
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+		pvc := &v1.PersistentVolumeClaim{}
+		if err := p.kubeClient.Get(ctx, types.NamespacedName{Name: volume.PersistentVolumeClaim.ClaimName, Namespace: pod.Namespace}, pvc); err != nil {
+			return fmt.Errorf("getting persistent volume claim %s, %w", volume.PersistentVolumeClaim.ClaimName, err)
+		}
+		pvc.Annotations["volume.kubernetes.io/selected-node"] = node.Name
```
Comment (wongma7) [inline pkg/controllers/provisioning/provisioner.go:262]: I think it is necessary for some routine to wait for every `PVC provision` to succeed/fail after setting this annotation to trigger `PVC provision`. Step 3 here: https://github.com/kubernetes/design-proposals-archive/blob/main/storage/volume-topology-scheduling.md#bind.

Because if a `PVC provision` fails then the CSI driver's provisioner controller signals failure by removing the selected-node annotation https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/a35bec6eb688ef7a0e31b340ed77bdb84010ace7/controller/controller.go#L1298. And some routine has to notice that + add the annotation again to trigger another `PVC provision`.

The default scheduler prebind implementation of this wait routine is: https://github.com/kubernetes/kubernetes/blob/39c76ba2edeadb84a115cc3fbd9204a2177f1c28/pkg/scheduler/framework/plugins/volumebinding/binder.go#L462-L465

**Comment by geoffcline** [comment]
👀👏

File: pkg/controllers/provisioning/provisioner.go (hunk 1 above)
Comment (ellistarn) [reply pkg/controllers/provisioning/provisioner.go:262]: Interesting. I may just implement a PVC scheduler that watches PVCs and pods and makes sure that the annotation is set for the node. Thanks for pointing this out.

File: pkg/controllers/persistentvolumeclaim/controller.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+		return reconcile.Result{}, nil
+	}
+	if nodeName, ok := pvc.Annotations[SelectedNodeAnnotation]; ok && nodeName == pod.Spec.NodeName {
+		return reconcile.Result{}, nil
+	}
+	if !c.isBindable(pod) {
+		return reconcile.Result{}, nil
+	}
+	pvc.Annotations = functional.UnionStringMaps(pvc.Annotations, map[string]string{SelectedNodeAnnotation: pod.Spec.NodeName})
+	if err := c.kubeClient.Update(ctx, pvc); err != nil {
```
Comment (suket22) [inline pkg/controllers/persistentvolumeclaim/controller.go:89]: I think I'm missing something obvious but why is this necessary? Once we bind the pod with the PVC volume to that node, won't this happen anyway?

File: pkg/controllers/selection/controller.go [hunk 3]
```diff
@@ -39,17 +39,19 @@ const controllerName = "selection"
...
 // NewController constructs a controller instance
 func NewController(kubeClient client.Client, provisioners *provisioning.Controller) *Controller {
 	return &Controller{
-		kubeClient:   kubeClient,
-		provisioners: provisioners,
-		preferences:  NewPreferences(),
+		kubeClient:     kubeClient,
+		provisioners:   provisioners,
+		preferences:    NewPreferences(),
+		volumeTopology: NewVolumeTopology(kubeClient),
```
Comment (suket22) [inline pkg/controllers/selection/controller.go:54]: Should we wrap VolumeTopology into Preferences and not expose it to the Controller?

File: pkg/controllers/persistentvolumeclaim/controller.go (hunk 2 above)
Comment (ellistarn) [reply pkg/controllers/persistentvolumeclaim/controller.go:89]: No! I thought that the PVC controller should be responsible for this, but it turns out that it's a prebind implementation within the kube scheduler. Any scheduler implementation (i.e. karpenter) is responsible for also doing volume binding. 

This is unintuitive to me, and I feel like the kvc or pvc controller (or kubescheduler, for that matter) should be responsible for this. This was brought up by @wongma7 in a previous comment.

File: pkg/controllers/selection/controller.go (hunk 3 above)
Comment (ellistarn) [reply pkg/controllers/selection/controller.go:54]: I think these are fully separate concepts. Preferences is all about relaxing preferred node affinities. Volume topology is a fully separate scheduling plugin.

Comment (ellistarn) [reply pkg/controllers/selection/controller.go:54]: That said, as we continue to build out scheduling features, I'm always game for a reimagining of the implementation to something more streamlined -- there just wasn't anything obvious to me as I've built it piece by piece.

Comment (suket22) [reply pkg/controllers/selection/controller.go:54]: > Preferences is all about relaxing preferred node affinities.

Aaah okay, that makes sense - on first reading I thought Preferences was the way we're getting all the scheduling logic into the controller. I suppose it was just first. 

Agreed though, it might be nice to have some kind of uniform contract for all our scheduler plugins to implement and unifying all of them into a different wrapper maybe?

File: pkg/controllers/persistentvolumeclaim/controller.go (hunk 2 above)
Comment (suket22) [reply pkg/controllers/persistentvolumeclaim/controller.go:89]: TIL!

**Review by suket22 — APPROVED** [review]

File: pkg/test/nodes.go [hunk 4]
```diff
@@ -16,25 +16,20 @@ package test
...
 	"github.com/imdario/mergo"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 type NodeOptions struct {
-	Name          string
-	Labels        map[string]string
-	Annotations   map[string]string
+	metav1.ObjectMeta
```
Comment (suket22) [inline pkg/test/nodes.go:26]: This was a nice refactor 💯
